docs/design/extending-api.md

```diff
@@ -0,0 +1,169 @@
+# Extending the Kubernetes API
+
+This document describes the design for implementing custom API types in the Kubernetes API Server.
+
+
+## Object Model
+
+### The Schema Object
+The `Schema` object describes the object schema(s) of multiple versions of a custom object that the user wants to add
+to the Kubernetes API.
+
+Each Schema object has the standard Kubernetes metadata.
+
+Additionally it has an array of one more `Version` objects.
+
+### The Version Object
+The `Version` object describes a single concrete version of a custom object.  The `Version` object has
+two fields:
+   * `Name` which is the name of the version (e.g. `v1beta3`)
+   * `Spec` which is the specification of the version
+
+### The SchemaSpec Object
+The `SchemaSpec` object describes a concrete schema for an object, it consists of one or more `Field` objects.
```

Comment:
I read this, but I don't understand what the system would do with the schema. We don't perform server-side validation now, and all our documentation and client-side validation is based on swagger. If we were going to do schema validation, we'd need required/optional information. If we were going to provide documentation, we'd need descriptions.

Something also has to do more detailed field validation. I'd like to attach more standard validation properties (min, max, regexp, etc.) to the swagger schema, where possible to express validation that way. But that would also mean we'd need to represent that here.